File: python/data_structures/linkedlist/circular_linked_list/circular_linked_list.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None


class CircularLinkedList:
    def __init__(self):
        self.tail = None   # tail.next = head
        self.size = 0

    def is_empty(self):
        return self.tail is None

    def __len__(self):
        return self.size

    
    def add_first(self, value):
        new_node = Node(value)

        if self.is_empty():
            self.tail = new_node
            new_node.next = new_node     # circular: points to itself
        else:
            new_node.next = self.tail.next   # current head
            self.tail.next = new_node        # new_node becomes new head

        self.size += 1

    def add_last(self, value):
        self.add_first(value)
        self.tail = self.tail.next  # move tail to the new node

    def remove_first(self):
        if self.is_empty():
            raise IndexError("remove_first from empty list")

        head = self.tail.next
        value = head.value

        # Only one node
        if self.tail is head:
            self.tail = None
        else:
            self.tail.next = head.next  # bypass the old head

        self.size -= 1
        return value

    def get_first(self):
        if self.is_empty():
            raise IndexError("get_first from empty list")
        return self.tail.next.value  # head

    def get_last(self):
        if self.is_empty():
            raise IndexError("get_last from empty list")
        return self.tail.value

    def find(self, value):
        if self.is_empty():
            return None

        current = self.tail.next  # head

        while True:
            if current.value == value:
                return current
            current = current.next

            if current == self.tail.next:
                break  # completed full cycle

        return None
    
    def __repr__(self):
        if self.is_empty():
            return "CircularLinkedList([])"

        elements = []
        current = self.tail.next  # head

        while True:
            elements.append(repr(current.value))
            current = current.next
            if current == self.tail.next:
                break

        return f"CircularLinkedList([{', '.join(elements)}])"

File: java/data_structures/linkedlist/circular_linked_list/CircularLinkedList.java
package data_structures.linkedlist.circular_linked_list;

public class CircularLinkedList<T> {

    public static class Node<T> {
        T value;
        Node<T> next;

        Node(T value) {
            this.value = value;
        }
    }

    public Node<T> tail;  // tail.next = head
    public int size;

    public CircularLinkedList() {
        this.tail = null;
        this.size = 0;
    }


    public boolean isEmpty() {
        return tail == null;
    }

    public int size() {
        return size;
    }

    // Insert at beginning (head)
    public void addFirst(T value) {
        Node<T> newNode = new Node<>(value);

        if (isEmpty()) {
            tail = newNode;
            newNode.next = newNode;   // circular: points to itself
        } else {
            newNode.next = tail.next; // current head
            tail.next = newNode;      // newNode becomes new head
        }

        size++;
    }

    // Insert at end (tail)
    public void addLast(T value) {
        addFirst(value);
        tail = tail.next; // move tail to the newly added node
    }

    // Remove first (head)
    public T removeFirst() {
        if (isEmpty()) {
            throw new IllegalStateException("removeFirst from empty list");
        }

        Node<T> head = tail.next;
        T value = head.value;

        // Only one node in the list
        if (tail == head) {
            tail = null;
        } else {
            tail.next = head.next; // bypass old head
        }

        size--;
        return value;
    }

    // Get first element
    public T getFirst() {
        if (isEmpty()) {
            throw new IllegalStateException("getFirst from empty list");
        }
        return tail.next.value; // head
    }

    // Get last element
    public T getLast() {
        if (isEmpty()) {
            throw new IllegalStateException("getLast from empty list");
        }
        return tail.value;
    }

    // Find value in the cycle
    public Node<T> find(T value) {
        if (isEmpty()) return null;

        Node<T> current = tail.next; // head

        while (true) {
            if ((current.value == null && value == null) ||
                (current.value != null && current.value.equals(value))) {
                return current;
            }

            current = current.next;

            if (current == tail.next) {
                break; // full cycle completed
            }
        }

        return null;
    }

    @Override
    public String toString() {
        if (isEmpty()) {
            return "CircularLinkedList([])";
        }

        StringBuilder sb = new StringBuilder("CircularLinkedList([");
        Node<T> current = tail.next; // head

        while (true) {
            sb.append(current.value);
            current = current.next;

            if (current == tail.next) {
                break;
            }
            sb.append(", ");
        }

        sb.append("])");
        return sb.toString();
    }
}
